File: androidweekview/src/main/java/com/alamkanak/weekview/DateUtils.java
package com.alamkanak.weekview;

import android.content.Context;
import android.text.format.DateFormat;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

/**
 * Created by jesse on 6/02/2016.
 */
class DateUtils {
	@SuppressWarnings("SameParameterValue")
	static List<Calendar> getDateRange(int daysSinceToday, int size, int weekStart, int weekEnd) {
		final List<Calendar> days = new ArrayList<>();
		Calendar day;

		int today = today().get(Calendar.DAY_OF_WEEK);
		if (today > weekEnd) today -= 7;
		int offset = (today < weekStart) ? weekStart - today : 0;
		for (int dayNumber = daysSinceToday; days.size() <= size; dayNumber++) {
			day = today();
			day.add(Calendar.DATE, dayNumber - 1 + offset);

			if (day.get(Calendar.DAY_OF_WEEK) >= weekStart && day.get(Calendar.DAY_OF_WEEK) <= weekEnd)
				days.add(day);
		}

		return days;
	}

	static Calendar withTimeAtStartOfDay(Calendar date) {
		date.set(Calendar.HOUR_OF_DAY, 0);
		date.set(Calendar.MINUTE, 0);
		date.set(Calendar.SECOND, 0);
		date.set(Calendar.MILLISECOND, 0);
		return date;
	}

	static Calendar withTimeAtEndOfDay(Calendar date) {
		date.set(Calendar.HOUR_OF_DAY, 23);
		date.set(Calendar.MINUTE, 59);
		date.set(Calendar.SECOND, 59);
		date.set(Calendar.MILLISECOND, 999);
		return date;
	}

	static int getDaysUntilDate(Calendar date) {
		final long dateInMillis = date.getTimeInMillis();
		final long todayInMillis = today().getTimeInMillis();
		final long diff = dateInMillis - todayInMillis;
		return (int) (diff / Constants.DAY_IN_MILLIS);
	}

	static int getDisplayedDays(Calendar startDay, int size, int weekStart, int weekEnd) {
		int startDayIndex = startDay.get(Calendar.DAY_OF_WEEK);
		if (startDayIndex > weekEnd) startDayIndex -= 7; // TODO: Is this line correct?
		int offsetForWeekStart = (startDayIndex > weekStart) ? startDayIndex - weekStart : 0;

		int days = 0;

		for (int i = 0; i < Math.abs(size); i++) {
			startDay.add(Calendar.DATE, size > 0 ? 1 : -1);
			if (startDay.get(Calendar.DAY_OF_WEEK) >= weekStart && startDay.get(Calendar.DAY_OF_WEEK) <= weekEnd)
				days += size > 0 ? 1 : -1;
		}

		return days + offsetForWeekStart;
	}

	/**
	 * Checks if two hourLines are on the same day.
	 *
	 * @param dayOne The first day.
	 * @param dayTwo The second day.
	 * @return Whether the hourLines are on the same day.
	 */
	static boolean isSameDay(Calendar dayOne, Calendar dayTwo) {
		return dayOne.get(Calendar.YEAR) == dayTwo.get(Calendar.YEAR)
				&& dayOne.get(Calendar.DAY_OF_YEAR) == dayTwo.get(Calendar.DAY_OF_YEAR);
	}


	/**
	 * Returns a calendar instance at the start of this day
	 *
	 * @return the calendar instance
	 */
	static Calendar today() {
		return today(0);
	}

	/**
	 * Returns a calendar instance at the start of this day
	 *
	 * @param offset the days to add to today
	 * @return the calendar instance
	 */
	static Calendar today(int offset) {
		final Calendar today = Calendar.getInstance();
		today.set(Calendar.HOUR_OF_DAY, 0);
		today.set(Calendar.MINUTE, 0);
		today.set(Calendar.SECOND, 0);
		today.set(Calendar.MILLISECOND, 0);
		today.add(Calendar.DATE, offset);
		return today;
	}

	static SimpleDateFormat getTimeFormat(Context context) {
		return DateFormat.is24HourFormat(context)
				? new SimpleDateFormat("H:mm", Locale.getDefault())
				: new SimpleDateFormat("h:mm a", Locale.getDefault());
	}
}

File: androidweekview/src/main/java/com/alamkanak/weekview/DateUtils.kt
package com.alamkanak.weekview

import android.content.Context
import android.text.format.DateFormat
import org.joda.time.DateTime
import org.joda.time.DateTimeConstants
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.ArrayList

internal object DateUtils {
	/**
	 * Returns a list of days starting at [startDay] with a length of [size] taking into account
	 * the week start day and length to exclude all weekend days.
	 *
	 * @param startDay the first day that should be in the returned list
	 * @param size The number of days to be returned
	 * @param weekStart The day index of the first day of the visible week. Example: [DateTimeConstants.MONDAY]
	 * @param weekLength The length of the displayed week. Example: `5` for a week from [DateTimeConstants.MONDAY] until [DateTimeConstants.FRIDAY]
	 *
	 * @return A list with all days that are within the visible week, starting with [startDay]
	 */
	fun getDateRange(startDay: DateTime, size: Int, weekStart: Int, weekLength: Int): List<DateTime> {
		val days = ArrayList<DateTime>()
		var day: DateTime
		var dayNumber = 0

		while (days.size <= size) {
			day = startDay.plusDays(dayNumber)

			if (day.dayOfWeek in weekStart until weekStart + weekLength)
				days.add(day)
			dayNumber++
		}

		return days
	}

	/**
	 * Calculates the offset of a day relative to the specified week start.
	 *
	 * If the day is not within the specified week start and end dates, an offset of `0` is returned instead.
	 *
	 * @param day The day to calculate the offset for
	 * @param weekStart The day index of the first day of the visible week. Example: [DateTimeConstants.MONDAY]
	 * @param weekLength The length of the displayed week. Example: `5` for a week from [DateTimeConstants.MONDAY] until [DateTimeConstants.FRIDAY]
	 *
	 * @return The offset of [day] relative to the specified [weekStart]. Never negative.
	 */
	fun offsetInWeek(day: DateTime, weekStart: Int, weekLength: Int): Int {
		val offset = day.dayOfWeek - weekStart
		return if (offset in 0 until weekStart + weekLength - 1) offset else 0
	}

	/**
	 * Converts between from displayed days to actual days, accounting for skipped days (if [weekLength] < `7`).
	 *
	 * @param displayedDays The amount of displayed days, starting at the first visible day of the week.
	 * @param weekLength The length of the displayed week. Example: `5` for a week from [DateTimeConstants.MONDAY] until [DateTimeConstants.FRIDAY]
	 *
	 * @return The amount of actual days.
	 */
	fun actualDays(displayedDays: Int, weekLength: Int): Int {
		val skippedDays = if (displayedDays < 0)
			(displayedDays + 1) / weekLength * (7 - weekLength) - (7 - weekLength)
		else
			displayedDays / weekLength * (7 - weekLength)

		return displayedDays + skippedDays
	}

	/**
	 * Converts from actual days to displayed days, accounting for skipped days (if [weekLength] < `7`).
	 *
	 * @param actualDays The amount of actual days, starting at the first visible day of the week.
	 * @param weekLength The length of the displayed week. Example: `5` for a week from [DateTimeConstants.MONDAY] until [DateTimeConstants.FRIDAY]
	 *
	 * @return The amount of actual days.
	 */
	fun displayedDays(actualDays: Int, weekLength: Int): Int {
		val skippedDays = if (actualDays < 0)
			(actualDays + 1) / 7 * (7 - weekLength) - (7 - weekLength)
		else
			actualDays / 7 * (7 - weekLength)

		return actualDays - skippedDays
	}

	fun isSameDay(dayOne: DateTime, dayTwo: DateTime) = dayOne.year == dayTwo.year && dayOne.dayOfYear == dayTwo.dayOfYear

	fun getTimeFormat(context: Context): SimpleDateFormat {
		return if (DateFormat.is24HourFormat(context))
			SimpleDateFormat("H:mm", Locale.getDefault())
		else
			SimpleDateFormat("h:mm a", Locale.getDefault())
	}
}
